SignalServiceKit/src/Network/OutageDetection.swift

```diff
@@ -0,0 +1,122 @@
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            ensureCheckTimer()
+        }
+    }
+
+    // We only show the outage warning when we're certain there's an outage.
+    // DNS lookup failures, etc. are not considered an outage.
+    private func checkForOutageSync() -> Bool {
+        let host = CFHostCreateWithName(nil, "uptime.signal.org" as CFString).takeRetainedValue()
+        CFHostStartInfoResolution(host, .addresses, nil)
+        var success: DarwinBoolean = false
+        guard let addresses = CFHostGetAddressing(host, &success)?.takeUnretainedValue() as NSArray? else {
+            Logger.error("\(logTag) CFHostGetAddressing failed: no addresses.")
+            return false
+        }
+        guard success.boolValue else {
+            Logger.error("\(logTag) CFHostGetAddressing failed.")
+            return false
+        }
+        var isOutageDetected = false
+        for case let address as NSData in addresses {
+            var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
+            if getnameinfo(address.bytes.assumingMemoryBound(to: sockaddr.self), socklen_t(address.length),
+                           &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST) == 0 {
+                let addressString = String(cString: hostname)
+                let kHealthyAddress = "127.0.0.1"
+                let kOutageAddress = "127.0.0.2"
+                if addressString == kHealthyAddress {
+                    // Do nothing.
+                } else if addressString == kOutageAddress {
+                    isOutageDetected = true
+                } else {
+                    owsFail("\(logTag) unexpected address: \(addressString)")
+                }
+            }
+        }
+        return isOutageDetected
+    }
+
+    private func checkForOutageAsync() {
+        DispatchQueue.global().async {
+            let isOutageDetected = self.checkForOutageSync()
+            DispatchQueue.main.async {
+                self.hasOutage = isOutageDetected
+            }
+        }
+    }
+
+    private var checkTimer: Timer?
+    private func ensureCheckTimer() {
+        // Only monitor for outages in the main app.
+        guard CurrentAppContext().isMainApp else {
+            return
+        }
+
+        if mayHaveOutage {
+            if checkTimer != nil {
+                // Already has timer.
+                return
+            }
+
+            // The TTL of the DNS record is 60 seconds.
+            checkTimer = WeakTimer.scheduledTimer(timeInterval: 60, target: self, userInfo: nil, repeats: true) { [weak self] _ in
+                SwiftAssertIsOnMainThread(#function)
+
+                guard CurrentAppContext().isMainAppAndActive else {
+                    return
+                }
+
+                guard let strongSelf = self else {
+                    return
+                }
+
+                strongSelf.checkForOutageAsync()
+            }
+        } else {
```

Comment:
Why do we disable the uptime monitor when there's been a network failure? 

Won't there often be network failures during the very service interruptions we want to advertise?

nit: and if we do want this check, a guard would be more legible than this short dangling else.

Reply:
Whoops, I inverted the semantics of these two methods while cleaning up.  I've renamed some of the methods and properties to hopefully make it clearer.  

We _only_ want to check for outages while experiencing failures connecting to the service - and want to stop checking as soon as we successfully connect.